forest/src/lib.rs

```diff
@@ -1,12 +1,605 @@
+    }
+
+    fn get(&self, node: Node<D, L>) -> &NodeContents<D, L> {
+        match self.arena.get(node.0) {
+            None => panic!("Forest - node has been deleted!"),
+            Some(node) => node,
+        }
+    }
+
+    fn get_mut(&mut self, node: Node<D, L>) -> &mut NodeContents<D, L> {
+        match self.arena.get_mut(node.0) {
+            None => panic!("Forest - node has been deleted (mut)!"),
+            Some(node) => node,
+        }
+    }
+
+    fn leaf(&self, node: Node<D, L>) -> &L {
+        match &self.get(node).leaf_or_children {
+            Leaf(leaf) => leaf,
+            Children(_) => panic!("Forest - branch nodes do not have leaf data"),
+        }
+    }
+
+    fn leaf_mut(&mut self, node: Node<D, L>) -> &mut L {
+        match &mut self.get_mut(node).leaf_or_children {
+            Leaf(leaf) => leaf,
+            Children(_) => panic!("Forest - branch nodes do not have leaf data (mut)"),
+        }
+    }
+
+    fn children(&self, node: Node<D, L>) -> &[Node<D, L>] {
+        match &self.get(node).leaf_or_children {
+            Leaf(_) => panic!("Forest - leaf nodes do not have children"),
+            Children(children) => children,
+        }
+    }
+
+    fn children_mut(&mut self, node: Node<D, L>) -> &mut Vec<Node<D, L>> {
+        match &mut self.get_mut(node).leaf_or_children {
+            Leaf(_) => panic!("Forest - leaf nodes do not have children (mut)"),
+            Children(children) => children,
+        }
+    }
+
+    fn siblings(&self, node: Node<D, L>) -> &[Node<D, L>] {
+        let parent = self.get(node).parent;
+        match parent {
+            None => &self.roots,
+            Some(parent) => self.children(parent),
+        }
+    }
+
+    fn siblings_mut(&mut self, node: Node<D, L>) -> &mut Vec<Node<D, L>> {
+        let parent = self.get(node).parent;
+        match parent {
+            None => &mut self.roots,
+            Some(parent) => self.children_mut(parent),
+        }
+    }
+
+    fn sibling_index(&self, node: Node<D, L>) -> usize {
+        self.siblings(node)
+            .iter()
+            .position(|t| *t == node)
+            .expect("Forest - missing child")
+    }
+}
+
+impl<D, L> Node<D, L> {
+    fn new(index: Index) -> Node<D, L> {
+        Node(index, PhantomData)
+    }
+
+    /// Check if this node is still valid (has not been deleted).
+    pub fn is_valid(self, f: &Forest<D, L>) -> bool {
+        f.arena.contains(self.0)
+    }
+
+    /// Get this node's parent. Returns `None` if already at the root.
+    pub fn parent(self, f: &Forest<D, L>) -> Option<Node<D, L>> {
+        f.get(self).parent
+    }
+
+    /// Get the root of the tree containing this node. (This is the same as
+    /// calling `parent()` repeatedly.)
+    pub fn root(self, f: &Forest<D, L>) -> Node<D, L> {
+        let mut root = self;
+        while let Some(parent) = root.parent(f) {
+            root = parent;
+        }
+        root
+    }
+
+    /// Get the data for this node.
```

Comment:
Consider referring to this as "node data". Not sure if worth bothering. But "leaf data" should never be called just "data", at least.